d3a_interface/sim_results/kpi.py

```diff
@@ -145,11 +145,74 @@ def update_area_kpi(self, area_dict, core_stats):
         self._accumulate_energy_trace(core_stats)
 
 
+class SavingsKPI:
+    def __init__(self):
+        self.producer_ess_set = set()
+        self.consumer_ess_set = set()
+        self.house_list = list()
+        self.fit_revenue = 0.
+        self.utility_bill = 0.
+        self.base_case_revenue = 0.
+        self.d3a_revenue = 0.
+        self.saving_absolute = 0.
+        self.saving_percentage = 0.
+
+    def calculate_saving_kpi(self, area_dict, core_stats, gf_alp):
+        """
+        Args:
+            area_dict: contain nested area info
+            core_stats: contain area's raw/key statistics
+            gf_alp: grid_fee_along_the_path - cumulative grid fee from root to target area
+
+        Returns:
+        """
+        self.populate_area_type(area_dict)
+
+        # fir_excl_gf_alp: feed-in tariff rate excluding grid fee along path
+        fir_excl_gf_alp = self.get_feed_in_tariff_rate_excluding_path_grid_fees(
+            core_stats.get(area_dict["uuid"], {}), gf_alp)
+        # mmr_incl_gf_alp: market maker rate include grid fee along path
+        mmr_incl_gf_alp = self.get_market_maker_rate_including_path_grid_fees(
+            core_stats.get(area_dict["uuid"], {}), gf_alp)
+        for trade in core_stats.get(area_dict["uuid"], {}).get("trades", []):
+            if trade["seller_origin_id"] in self.producer_ess_set:
+                self.fit_revenue += fir_excl_gf_alp * trade["energy"]
+                self.d3a_revenue += trade["price"]
+            if trade["buyer_origin_id"] in self.consumer_ess_set and \
+                    trade["seller_origin_id"] not in self.producer_ess_set:
+                self.utility_bill += mmr_incl_gf_alp * trade["energy"]
+                self.d3a_revenue -= trade["price"]
+        self.base_case_revenue = self.fit_revenue - self.utility_bill
+        self.saving_absolute = self.d3a_revenue - self.base_case_revenue
+        self.saving_percentage = ((self.saving_absolute / self.base_case_revenue) * 100
```

Comment:
Although I got a hold of what these variables represent, it would be great if you could either add a comment in `__init__` for each of them or add a description in the docstring. Thanks a lot!

Reply:
Added: https://github.com/gridsingularity/d3a-interface/pull/223/files#diff-4fe53216bd5f83d932a02c99f5062b7714e165902d3aad6274ef33e009b5b2cbR151-R158